File: desktop/scripts/chunk_parity.py
"""Print the chunk boundaries the *Python* app would pick, for diffing.

Counterpart to `src-tauri/examples/chunk_parity.rs`. Run both on the same file
and diff: identical output means the Rust decode + split port is faithful.

    .venv/bin/python desktop/scripts/chunk_parity.py data/some.mp3 4
"""
import sys
from pathlib import Path

# The Python app lives at the repo root, two levels up from this script.
sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

import numpy as np  # noqa: E402

from chunking import (  # noqa: E402
    DEFAULT_CHUNK_SECONDS,
    MIN_TAIL_SECONDS,
    SAMPLE_RATE,
    SPLIT_SEARCH_SECONDS,
    audio_duration,
    decode_range,
    find_split_index,
)


def main(argv: list[str]) -> int:
    if not argv:
        print("usage: chunk_parity.py <audio> [max_chunks]")
        return 2
    source = Path(argv[0])
    max_chunks = int(argv[1]) if len(argv) > 1 else 4

    duration = audio_duration(source)
    print(f"duration_sec={duration:.3f}" if duration else "duration_sec=none")

    start = 0.0
    for index in range(max_chunks):
        if duration is not None and start >= duration - MIN_TAIL_SECONDS:
            break
        audio = decode_range(source, start, DEFAULT_CHUNK_SECONDS + SPLIT_SEARCH_SECONDS)
        if audio.size <= int(MIN_TAIL_SECONDS * SAMPLE_RATE):
            break

        nominal_end = int(DEFAULT_CHUNK_SECONDS * SAMPLE_RATE)
        if audio.size > nominal_end:
            cut = min(find_split_index(audio, nominal_end), audio.size)
        else:
            cut = audio.size
        nxt = start + cut / SAMPLE_RATE

        checksum = float(np.abs(audio[:cut]).astype(np.float64).sum())
        print(
            f"chunk={index} start={start:.3f} decoded={audio.size} "
            f"cut={cut} next={nxt:.3f} sum={checksum:.6f}"
        )
        start = nxt
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: desktop/src-tauri/examples/chunk_parity.rs
//! Prints the chunk boundaries this crate would pick for an audio file.
//!
//! Exists to be diffed against `scripts/chunk_parity.py`, which prints the same
//! thing from the Python app's `chunking` module. Identical output means the
//! decode + split port is faithful, which is the precondition for an A/B
//! transcript comparison telling us anything about the *model*.
//!
//!     cargo run --example chunk_parity -- ../../data/some.mp3 4

use std::path::PathBuf;

use ai_meeting_lib::chunking::decode::{audio_duration, decode_range};
use ai_meeting_lib::chunking::split::find_split_index;
use ai_meeting_lib::SAMPLE_RATE;

const CHUNK_SECONDS: f64 = 300.0;
const SPLIT_SEARCH_SECONDS: f64 = 20.0;
const MIN_TAIL_SECONDS: f64 = 0.2;

fn main() -> anyhow::Result<()> {
    let mut args = std::env::args().skip(1);
    let source = PathBuf::from(args.next().expect("usage: chunk_parity <audio> [max_chunks]"));
    let max_chunks: usize = args
        .next()
        .map(|value| value.parse().expect("max_chunks must be a number"))
        .unwrap_or(4);

    let duration = audio_duration(&source);
    println!("duration_sec={}", fmt_opt(duration));

    let mut start = 0.0f64;
    for index in 0..max_chunks {
        if let Some(total) = duration {
            if start >= total - MIN_TAIL_SECONDS {
                break;
            }
        }
        let audio = decode_range(&source, start, CHUNK_SECONDS + SPLIT_SEARCH_SECONDS)?;
        if audio.len() <= (MIN_TAIL_SECONDS * SAMPLE_RATE as f64) as usize {
            break;
        }

        let nominal_end = (CHUNK_SECONDS * SAMPLE_RATE as f64) as usize;
        let cut = if audio.len() > nominal_end {
            find_split_index(&audio, nominal_end).min(audio.len())
        } else {
            audio.len()
        };
        let next = start + cut as f64 / SAMPLE_RATE as f64;

        // The checksum catches a decode that lines up on length but not on
        // samples, which is exactly the failure a resampler difference causes.
        println!(
            "chunk={index} start={start:.3} decoded={} cut={cut} next={next:.3} sum={:.6}",
            audio.len(),
            audio[..cut].iter().map(|s| s.abs() as f64).sum::<f64>()
        );
        start = next;
    }
    Ok(())
}

fn fmt_opt(value: Option<f64>) -> String {
    value.map(|v| format!("{v:.3}")).unwrap_or_else(|| "none".into())
}
